src/transforms/kubernetes/mod.rs

```diff
@@ -103,108 +143,103 @@ enum BuildError {
+        // defined max_retry_timeout.
+        let mut retry_timeout = self.max_retry_timeout;
         loop {
+            // Wait for bit before trying to watch again.
+            // We are sampling from a uniform distribution here.
+            let timeout = rand::thread_rng().gen_range(Duration::default(), retry_timeout);
+            let _ = Delay::new(Instant::now() + timeout)
+                .compat()
+                .await
+                .expect("Timer not set.");
+
             // Build watcher stream
-            let mut watcher = match self.client.watch_metadata(version.clone(), error.take()) {
-                Ok(watcher) => watcher,
-                Err(error) => return BuildError::WatchStreamBuild { source: error },
-            };
+            let mut watcher = self
+                .client
+                .watch_metadata(version.clone(), error.take())
+                .context(WatchStreamBuild)?
+                .compat();
             info!("Watching Pod metadata.");
 
             // Watch loop
-            let runtime_error = loop {
-                break match watcher.into_future().compat().await {
-                    Ok((Some(event), tail)) => {
-                        watcher = tail;
+            error = Some(RuntimeError::WatchUnexpectedlyEnded);
+            retry_timeout = self.max_retry_timeout.min(retry_timeout * 2);
 
+            while let Some(next) = watcher.next().await {
+                match next {
+                    Ok(event) => {
                         self.delete_update();
                         version = self.update(event).or(version);
 
                         self.metadata.refresh();
-                        continue;
                     }
-                    Ok((None, _)) => RuntimeError::WatchUnexpectedlyEnded,
-                    Err((err, _)) => err,
-                };
-            };
+                    Err(err) => {
+                        match err {
+                            // Keep the retry_timeout for errors that could
+                            // be caused by the api server being overloaded.
+                            RuntimeError::FailedConnecting { .. }
+                            | RuntimeError::ConnectionStatusNotOK { .. }
+                            | RuntimeError::WatchConnectionErrored { .. }
+                            | RuntimeError::WatchUnexpectedlyEnded => (),
+                            // Optimistically try the shortest timeout.
+                            _ => retry_timeout = Duration::from_secs(1),
+                        }
+                        error = Some(err);
+                        break;
+                    }
+                }
+            }
 
             warn!(
-                message = "Temporary stoped watching Pod metadata.",
-                reason = %runtime_error
+                message = "Temporarily stoped watching Pod metadata.",
+                reason = ?error
```

Comment:
```suggestion
                reason = %error
```

This should be display if it impls Error